### PR TITLE
Prices module

### PR DESCRIPTION
**Pull request by rawfalafel**

### CODE REVIEW

File: sdk/src/prices/calculatePoolPrices.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  );
+
+  const amountOutThreshold = new BN(
+    new Decimal(amount.toString())
+      .mul(new Decimal(pool.sqrtPrice.toString()).pow(2))
+      .div(priceImpactThreshold)
+      .toString()
+  );
+
+  return estimatedAmountOut.lt(amountOutThreshold);
```
Comment (yugure-orca) [inline sdk/src/prices/calculatePoolPrices.ts:122]: lt --> gte ?

File: sdk/src/prices/calculatePoolPrices.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  tickArrays: TickArray[],
+  isTokenA: boolean,
+  amount: BN,
+  priceImpactThreshold: number
+): boolean {
+  const { estimatedAmountOut } = swapQuoteWithParams(
+    {
+      whirlpoolData: pool,
+      aToB: isTokenA,
+      amountSpecifiedIsInput: true,
```
Comment (yugure-orca) [inline sdk/src/prices/calculatePoolPrices.ts:106]: To confirm that X token can be swapped for 1000 USDC,  ExactOut (AmountIsInput=false) mode is helpful.

Liquidity of meme tokens tends to be deeper in the direction of swap USDC for meme token, but the reverse direction, liquidity is small.

This is just a feeling...
BaseToQuote should be ExactOut
QuoteToBase should be ExactIn

**Review by yugure-orca — COMMENTED** [review]
I agree with the idea of trusting pools that can exchange tokens above a certain level with a price impact within a certain range.
Jupiter has similar criteria, so I believe that if pools have the ability to exchange about 1000 USDC, it will be arbitraged even if it illegally inflate the price.
https://docs.jup.ag/notes/getting-your-token-on-jupiter

The comments relate to three main points

**1. MintA, MintB and aToB direction**
orderMints is needed before calculating PDAUtil.getWhirlpool.
It will be a different PDA and the pool will not be discoverable.

**2. Using ExactOut**
The "token price" used to display the USDC price in the tokenPanel and assets should be backed by a liquidity depth that allows tokens to be sold and exchanged for USDC or SOL.

In that case, the liquidity deepness check would be to ensure that token to n SOL and token to n USDC trades do not increase the amount of token input.
In other words, it would be a verification by ExactOut to obtain n SOL and n USDC.

The reverse direction is valid as a buyer, but liquidity tends to be thicker in the direction of buying (meme) tokens, and I think the reverse is less likely.

**3. sqrtprice (Q64.Q64)**.
I think we need to use PriceMath to convert the sqrtPrice to price.

**Comment by rawfalafel** [comment]
> I agree with the idea of trusting pools that can exchange tokens above a certain level with a price impact within a certain range. Jupiter has similar criteria, so I believe that if pools have the ability to exchange about 1000 USDC, it will be arbitraged even if it illegally inflate the price. https://docs.jup.ag/notes/getting-your-token-on-jupiter
> 
> The comments relate to three main points
> 
> **1. MintA, MintB and aToB direction** orderMints is needed before calculating PDAUtil.getWhirlpool. It will be a different PDA and the pool will not be discoverable.
> 
> **2. Using ExactOut** The "token price" used to display the USDC price in the tokenPanel and assets should be backed by a liquidity depth that allows tokens to be sold and exchanged for USDC or SOL.
> 
> In that case, the liquidity deepness check would be to ensure that token to n SOL and token to n USDC trades do not increase the amount of token input. In other words, it would be a verification by ExactOut to obtain n SOL and n USDC.
> 
> The reverse direction is valid as a buyer, but liquidity tends to be thicker in the direction of buying (meme) tokens, and I think the reverse is less likely.
> 
> **3. sqrtprice (Q64.Q64)**. I think we need to use PriceMath to convert the sqrtPrice to price.

Thanks for the thoughtful review! I should have been more clear that the actual implementation was likely full of bugs, but I wanted feedback on the overall architecture and exported function interfaces :sweat_smile:

File: sdk/src/prices/fetchPoolPrices.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    if (!mintInfo) {
+      throw new Error(`Mint account does not exist: ${mints[index].toBase58()}`);
+    }
+
+    acc[mints[index].toBase58()] = mintInfo.decimals;
+    return acc;
+  }, {} as DecimalsMap);
+}
+
+export async function fetchPoolsForMints(
```
Comment (rawfalafel) [inline sdk/src/prices/fetchPoolPrices.ts:54]: TODO: I will modify this to be a more generalizable function to fetch pool accounts from a list of mint addresses

**Comment by rawfalafel** [comment]
would love to get some eyes on this. some things to consider: 
 - can this replace `computeWhirlpoolPrices` in `web`
 - can this replace `computeWhirlpoolPrices` in `server`
 - can the graph traversal logic be shared with the route traversal logic used elsewhere: https://app.asana.com/0/1200524542633404/1203400626933101
 - can some of this logic be reused to calculate prices for specific fee tiers for calculating price impact: https://app.asana.com/0/1202413563699470/1204095867950654

**Review by yugure-orca — COMMENTED** [review]
very helpful module! 😊

**Comment by odcheung** [comment]
Sharing some notes on my evaluation on the PriceModule architecture. Discussed with yutaro offline on questions.

## Module summary
- Provide a shared module to provide the source of truth of token prices based off of Whirlpool account data.
- Add additional liquidity criteria to filter out tokens with outlier prices

## Implementation Changes
- Replace `computeWhirlpoolPrices` in `web` & server
- In web, this deprecates the legacy pools from `useTokenPrices` & changes the implementation from deriving prices from pool-graph-traversal on all pool edges (hop-2) to one considering quote-token edges only

## Questions

Q. Do we need a replacement for `web` token price calculation? Since USD price is supplementary, non-critical information, can we solely retrieve this off of the server? There may be more efficient ways of calculating prices if this module is for the server alone.

A: 
- general design philosophy is `web` data should be able to be fetched from the chain (decentralized) with the server providing quick lookups to speed up the experience.
- There can be edge cases where server hasn't picked up on prices yet but the UI needs to present a price sooner (ex. WP creation flow)
- The price stream that is being built on server-side will also depend on this implementation

Q: Doing it by quote-token edges mean our team would have to maintain this list. Which means new pools w/ unknown quote-token will be missed. What are the trade offs to using all-edges w/ hop 2 for the graph traversal?

A:
- 2-hop via all edges is more computationally heavy, but can provide prices similar to our trade routes. However, since we want to avoid considering trade-amount when deciding trade routes, using quote-edges provide an accurate enough estimation on the price.
- As a business decision, LPs / pool creators should learn that in order to get the most liquidity, they need to provide a trading pair towards popular tokens (ex. SOL, USDC etc). If they are launching a new stable-coin, it should have a stable-coin/(SOL/USDC/etc) pair.

Q. The prices provided by this module (and current impl) is not accurate enough for some of our use-cases on web. 

For example, estimated execution price vs USD value on the trade module. User can get conflicting messages about the two if price module (quote-token based) differs too much from swap estimation (route / liquidity / trade-amount based)

Or deposit/withdrawal modules, the price is not supplemental information but it can be inaccurate to the actual converted value based on how big the user's position size is.

A: The UI should move away from USD-centric information and use it as supplemental information only. We need clarify on the module comments that there can be discrepancies between price module prices and execution price.

File: sdk/src/prices/calculatePoolPrices.ts [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+ *  - tickArrayMap: A map of pool addresses to tick array data
+ *  - decimalsMap: A map of token mint addresses to token decimals
+ *
+ * fetchPoolPrices.ts provides functions to fetch these accounts
+ */
+export function calculatePoolPrices(
+  mints: PublicKey[],
+  poolMap: PoolMap,
+  tickArrayMap: TickArrayMap,
+  decimalsMap: DecimalsMap,
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/prices/calculatePoolPrices.ts:42]: note that this namespace will can also host functions related to getting prices for a specific pool.

[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/prices/calculatePoolPrices.ts:42]: Another thing that's popping out to me is that we should be re-using pool-graph that's being used on smart-router to do the walk.

The UI already needs to generate this graph once to support routing. We can save some computation if we want to support prices on top of that.

File: sdk/src/prices/fetchPoolPrices.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      const bToATickArrayPublicKeys = SwapUtils.getTickArrayPublicKeys(
+        pool.tickCurrentIndex,
+        pool.tickSpacing,
+        false,
+        programId,
+        new PublicKey(poolAddress)
+      );
+
+      // Fetch tick arrays in both directions
+      return [...aToBTickArrayPublicKeys, ...bToATickArrayPublicKeys.slice(1)];
```
[the remark this answers is not in the document]
Comment (rawfalafel) [reply sdk/src/prices/fetchPoolPrices.ts:112]: ah, good catch!

File: sdk/tests/integration/get_pool_prices.test.ts [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+  getDefaultAquarium,
+  initTestPoolWithLiquidity,
+} from "../utils/init-utils";
+import * as assert from "assert";
+import { u64 } from "@solana/spl-token";
+import Decimal from "decimal.js";
+import { MathUtil } from "@orca-so/common-sdk";
+import { PublicKey } from "@solana/web3.js";
+
+describe.only("get_pool_prices", () => {
```
[the remark this answers is not in the document]
Comment (rawfalafel) [reply sdk/tests/integration/get_pool_prices.test.ts:25]: nope!

**Comment by odcheung** [comment]
This graph traversal requirements can be satisfied by the PoolGraph stack that we've already built for smart-router.

Quoting tokens via a set of quoteTokens is equivalent to finding the walks between token & mainQuoteToken via a set list of intermediate tokens (the rest of the quoteTokens).

We can ship this as is to unblock. However, I recommend that we switch it to use the PoolGraph impl during our implementation of multi-hop price impact. 
1. Multi-hop price impact can't use this traversal method
2. Reduce code to maintain
3. PoolGraph is already generated on our UI. We should re-use it instead of having another map in memory.

File: sdk/src/prices/calculatePoolPrices.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rawfalafel) [reply sdk/src/prices/calculatePoolPrices.ts]: > [side note for future] Instead of storing decimals only, can we use the types defined token package that @terranmoccasin is building? Having a mapping from token -> price can also be useful since users may want to present the info as "SYMBOL - PRICE" instead of "MINT - PRICE"

potentially! although IIRC, the token package will not just fetch decimal data but metadata from metaplex accounts and other sources. we may not want to require callers of `PriceModule` to also fetch all of this additional data just to calculate decimals

on the other hand, reusable code to fetch decimals that's shared by both the token package and the price module sounds like a reasonable idea

**Comment by rawfalafel** [comment]
Agreed! note that @matt-allan already created a task for this: https://app.asana.com/0/1200524542633404/1203400626933101

**Comment by rawfalafel** [comment]
addressed comments. feel free to re-organize the code further, if that is helpful

File: sdk/src/prices/calculatePoolPrices.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (terranmoccasin) [reply sdk/src/prices/calculatePoolPrices.ts]: > the token package will not just fetch decimal data but metadata from metaplex accounts and other sources.

Using the current `TokenFetcher` implementation, it's actually possible to fetch only the on-chain mint account data. You could create a `TokenFetcher` that doesn't have any `MetadataProvider`s added.

https://github.com/orca-so/orca-sdks/blob/main/packages/token-sdk/src/token-fetcher.ts#L40

Fetcher always grabs mint account data first before supplementing the returned `Token` object with optional metadata.

**Review by yugure-orca — APPROVED** [review]
LGTM🙂

File: sdk/src/prices/price-module.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    refresh = true
+  ) {
+    const poolMap = await PriceModule.fetchPoolsForMints(ctx, mints, config, refresh);
+    const tickArrayMap = await PriceModule.fetchTickArraysForPools(ctx, poolMap, config, refresh);
+    const decimalsMap = await PriceModule.fetchDecimalsForMints(ctx, mints, refresh);
+
+    return { poolMap, tickArrayMap, decimalsMap };
+  }
+
+  static async fetchPoolsForMints(
```
Comment (odcheung) [inline sdk/src/prices/price-module.ts:57]: Slight concern about this loop now that we realize deriving PDA is pretty slow. We are generating PDAs for all combinations of mints & tick-spacing which may be overkill.

If we are using the price module in useTokenPrices, we will need these fetches to be performant. Suggest we add another fetch function here to allow fetching pools w/ address directly.

**Review by odcheung — APPROVED** [review]
